### PR TITLE
Update C Interop Tutorial to make it platform independent

### LINKED ISSUE #2925: Update the C Interop Tutorial to make it platform independent
### Description

The tutorial follows [this](https://github.com/dart-lang/samples/tree/master/ffi/hello_world) example. While the `hello.dart` program in the example uses different pathnames to use platform-specific dynamic libraries, the tutorial shows the following line, which opens the dynamic library specific to macOS. 
```dart
final dylib = ffi.DynamicLibrary.open('hello_world.dylib');
```
To make the tutorial platform-independent, we may use the path for dynamic libraries for the current platform as done in the example.

The following are the relative paths for the dynamic libraries as per different platforms.
- macOS -   `hello_library/libhello.dylib`
- Linux - `hello_library/libhello.so`
- Windows - `hello_library/Debug/hello.dll`

### File Details

Filename: `src/_guides/libraries/c-interop.md`
Link to Tutorial: https://dart.dev/guides/libraries/c-interop

### PR DESCRIPTION
**Pull request by shubhanshu02**
Fixes #2925 
Added the steps to use the path of the dynamic library as per the platform the program is run on.

### CODE REVIEW

File: src/_guides/libraries/c-interop.md [hunk 1]
```diff
@@ -90,11 +90,13 @@ The [`hello.dart` file]({{ page.hw}}/hello.dart)
...
 1. Import dart:ffi.
-2. Create a typedef with the FFI type signature of the C function.
-3. Create a typedef for the variable that you'll use when calling the C function.
-4. Open the dynamic library that contains the C function.
-5. Get a reference to the C function, and put it into a variable.
-6. Call the C function.
+2. Import the path library that you'll use to store the path of dynamic library.
+3. Create a typedef with the FFI type signature of the C function.
+4. Create a typedef for the variable that you'll use when calling the C function.
+5. Create a variable to store the path of the dynamic library depending upon your platform.
```
Comment (kwalrath) [inline src/_guides/libraries/c-interop.md:96]:
```suggestion
5. Create a variable to store the path of the dynamic library.
```

File: src/_guides/libraries/c-interop.md [hunk 2]
```diff
@@ -90,11 +90,13 @@ The [`hello.dart` file]({{ page.hw}}/hello.dart)
 illustrates the steps for using dart:ffi to call a C function:
 
 1. Import dart:ffi.
-2. Create a typedef with the FFI type signature of the C function.
-3. Create a typedef for the variable that you'll use when calling the C function.
-4. Open the dynamic library that contains the C function.
-5. Get a reference to the C function, and put it into a variable.
-6. Call the C function.
+2. Import the path library that you'll use to store the path of dynamic library.
```
Comment (kwalrath) [inline src/_guides/libraries/c-interop.md:93]: If importing Platform, Directory, & path is really necessary, we should probably put that in the first step:

1. Import the dart:ffi library and APIs to construct the library path.

If we don't want to unconditionally recommend it (@dcharkes?), that makes the writing more complicated, but we should preface this with _Optional:_:

2. _Optional:_ Import APIs to construct the library path.

File: src/_guides/libraries/c-interop.md [hunk 3]
```diff
@@ -103,24 +105,41 @@ Here's the code for each step.
 import 'dart:ffi' as ffi;
 ```
 
-2. Create a typedef with the FFI type signature of the C function. <br>
+2. Import the path library that you'll use to store the path of dynamic library.
```
Comment (kwalrath) [inline src/_guides/libraries/c-interop.md:108]: See suggestion above

File: src/_guides/libraries/c-interop.md [hunk 4]
```diff
@@ -103,24 +105,41 @@ Here's the code for each step.
...
-3. Create a typedef for the variable that you'll use when calling the C function.
+4. Create a typedef for the variable that you'll use when calling
+   the C function.
 ```dart
 typedef HelloWorld = void Function();
 ```
 
-4. Open the dynamic library that contains the C function.
+5. Create a variable to store the path of the dynamic library 
+   depending upon your platform.
```
Comment (kwalrath) [inline src/_guides/libraries/c-interop.md:128]: See suggestion above.

File: src/_guides/libraries/c-interop.md [hunk 5]
```diff
@@ -103,24 +105,41 @@ Here's the code for each step.
...
+```dart
+  var libraryPath = path.join(Directory.current.path, 'hello_library', 'libhello.so');
+  if (Platform.isMacOS)
+    libraryPath = path.join(Directory.current.path, 'hello_library', 'libhello.dylib');
+  if (Platform.isWindows)
+    libraryPath = path.join(Directory.current.path, 'hello_library', 'Debug', 'hello.dll');
+```
+
+6. Open the dynamic library that contains the C function using its path.
```
Comment (kwalrath) [inline src/_guides/libraries/c-interop.md:137]:
```suggestion
6. Open the dynamic library that contains the C function.
```

**Review by kwalrath — COMMENTED** [review]
Thank you, @shubhanshu02! I've put in my 2 cents from a writer's perspective, but I'd like @dcharkes (or someone equally knowledgeable) to weigh in, as well.

**Comment by dcharkes** [comment]
lgtm.  This basically inlines the platform independent-ness from the first sample linked (`hello.dart`).

**Review by dcharkes — APPROVED** [review]

File: src/_guides/libraries/c-interop.md (hunk 2 above)
Comment (shubhanshu02) [reply src/_guides/libraries/c-interop.md:93]: @kwalrath, I am not sure if these APIs are really necessary. Though the example mentioned in the tutorial uses these to define the path variable for the dynamic library, but I could get the same result by passing the relative file address directly to the function like this:
```dart
final dylib = ffi.DynamicLibrary.open('hello_library/libhello.so');
```

File: src/_guides/libraries/c-interop.md [hunk 6]
```diff
@@ -103,24 +105,40 @@ Here's the code for each step.
...
 ```
 
-4. Open the dynamic library that contains the C function.
+5. Create a variable to store the path of the dynamic library.
+```dart
+  var libraryPath = path.join(Directory.current.path, 'hello_library', 'libhello.so');
+  if (Platform.isMacOS)
+    libraryPath = path.join(Directory.current.path, 'hello_library', 'libhello.dylib');
+  if (Platform.isWindows)
+    libraryPath = path.join(Directory.current.path, 'hello_library', 'Debug', 'hello.dll');
```
Comment (kwalrath) [inline src/_guides/libraries/c-interop.md:133]: These lines are really long; we need to shorten them. Also, I think we should follow the style guideline: https://dart.dev/guides/language/effective-dart/style#do-use-curly-braces-for-all-flow-control-structures. And maybe use an `else if`? So maybe this:

```suggestion
var libraryPath = path.join(Directory.current.path, 'hello_library',
    'libhello.so');
if (Platform.isMacOS) { 
  libraryPath = path.join(Directory.current.path, 'hello_library', 
      'libhello.dylib');
} else if (Platform.isWindows) { 
  libraryPath = path.join(Directory.current.path, 'hello_library', 
      'Debug', 'hello.dll');
} 
```

**Review by kwalrath — COMMENTED** [review]
Sorry for the very long delay!

LGTM but I think we might want to change the formatting of the code that does the platform-specific path construction. I've staged both versions. (Weirdly, initial spaces seem to be compressed... I don't know why.)

Staged with the PR as-is (go down to item 5):
https://kw-dartlang-3.firebaseapp.com/guides/libraries/c-interop#using-dartffi

Staged with my proposed change (again, item 5):
https://kw-www-dartlang-1.firebaseapp.com/guides/libraries/c-interop#using-dartffi

**Comment by kwalrath** [comment]
@dcharkes could you please check out my diffs at https://github.com/dart-lang/site-www/pull/2930#discussion_r612039135?

> Staged with the PR as-is (go down to item 5):
> https://kw-dartlang-3.firebaseapp.com/guides/libraries/c-interop#using-dartffi
> 
> Staged with my proposed change (again, item 5):
> https://kw-www-dartlang-1.firebaseapp.com/guides/libraries/c-interop#using-dartffi

File: src/_guides/libraries/c-interop.md (hunk 6 above)
Comment (dcharkes) [reply src/_guides/libraries/c-interop.md:133]: 👍

**Review by kwalrath — APPROVED** [review]

**Pull request merged**